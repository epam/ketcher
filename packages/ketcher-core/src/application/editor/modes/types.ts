export type LayoutMode =
  | 'flex-layout-mode'
  | 'snake-layout-mode'
  | 'sequence-layout-mode';
