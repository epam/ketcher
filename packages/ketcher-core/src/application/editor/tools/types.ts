export enum ToolName {
  monomer = 'monomer',
  preset = 'preset',
  selectRectangle = 'select-rectangle',
  bondSingle = 'bond-single',
  bondHydrogen = 'bond-hydrogen',
  erase = 'erase',
  clear = 'clear',
  hand = 'hand',
}
