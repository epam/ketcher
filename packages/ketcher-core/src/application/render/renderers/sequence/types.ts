import {
  ChemSequenceItemRenderer,
  EmptySequenceItemRenderer,
  NucleosideSequenceItemRenderer,
  NucleotideSequenceItemRenderer,
  PeptideSequenceItemRenderer,
  PhosphateSequenceItemRenderer,
  UnresolvedMonomerSequenceItemRenderer,
  UnsplitNucleotideSequenceItemRenderer,
} from 'application/render';
import { AmbiguousSequenceItemRenderer } from 'application/render/renderers/sequence/AmbiguousSequenceItemRenderer';

export type SequenceItemRenderer =
  | NucleotideSequenceItemRenderer
  | NucleosideSequenceItemRenderer
  | EmptySequenceItemRenderer
  | ChemSequenceItemRenderer
  | PeptideSequenceItemRenderer
  | PhosphateSequenceItemRenderer
  | UnresolvedMonomerSequenceItemRenderer
  | UnsplitNucleotideSequenceItemRenderer
  | AmbiguousSequenceItemRenderer;
