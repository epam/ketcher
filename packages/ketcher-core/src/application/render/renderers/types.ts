export type StructureBbox = {
  width: number;
  height: number;
  left: number;
  right: number;
  top: number;
  bottom: number;
};
