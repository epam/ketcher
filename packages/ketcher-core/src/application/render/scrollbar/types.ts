// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type RaphaelElement = Record<string, any>;

export type RaphaelRectAttr = {
  x: number;
  y: number;
  width: number;
  height: number;
  r: number;
};
