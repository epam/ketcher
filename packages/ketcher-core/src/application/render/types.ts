import { Selection } from 'd3';

export type D3SvgElementSelection<
  ElementType extends SVGElement,
  T,
> = Selection<ElementType, T, HTMLElement, never>;
