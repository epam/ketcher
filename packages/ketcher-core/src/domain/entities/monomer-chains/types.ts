import { Nucleoside } from 'domain/entities/Nucleoside';
import { Nucleotide } from 'domain/entities/Nucleotide';
import { MonomerSequenceNode } from 'domain/entities/MonomerSequenceNode';

export type SubChainNode = MonomerSequenceNode | Nucleoside | Nucleotide;
