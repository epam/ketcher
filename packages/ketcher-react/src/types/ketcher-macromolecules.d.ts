declare module 'ketcher-macromolecules' {
  import * as React from 'react';
  interface MacromoleculesEditorProps {
    ketcherId: string;
    togglerComponent?: JSX.Element;
  }
  const MacromoleculesEditor: React.ComponentType<MacromoleculesEditorProps>;
  export default MacromoleculesEditor;
}
