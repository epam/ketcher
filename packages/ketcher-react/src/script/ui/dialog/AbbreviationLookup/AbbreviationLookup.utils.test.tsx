/****************************************************************************
 * Copyright 2021 EPAM Systems
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 ***************************************************************************/

import { FilterOptionsState } from '@mui/material';
import {
  filterOptions,
  getOptionLabel,
  getSimilarity,
  highlightOptionLabel,
} from './AbbreviationLookup.utils';
import { AbbreviationOption } from './AbbreviationLookup.types';
import {
  createGenericOption,
  createOption,
} from './AbbreviationLookup.test.utils';

describe('AbbreviationLookup Utils', () => {
  describe('highlightOptionLabel', () => {
    const option = createGenericOption('Very long test name', 'SHORTABBR');

    it('Should return the label if there is not match with lookup value', () => {
      expect(highlightOptionLabel(option, 'no-match-string')).toBe(
        option.label,
      );
    });

    it('Should not highlight if lookup value length is short', () => {
      expect(highlightOptionLabel(option, 'l')).toMatchSnapshot();
    });

    it('Should return marked name', () => {
      expect(highlightOptionLabel(option, 'long')).toMatchSnapshot();
    });

    it('Should ignore space at the beginning and the end', () => {
      expect(highlightOptionLabel(option, '   HORT  ')).toMatchSnapshot();
    });

    it('Should return marked abbreviation', () => {
      expect(highlightOptionLabel(option, 'abbr')).toMatchSnapshot();
    });

    it('Should mark abbreviation because it shorter (bigger similarity)', () => {
      const newOption = createGenericOption(
        'ShortNameForTemplate',
        'SHORTABBR',
      );
      expect(highlightOptionLabel(newOption, 'short')).toMatchSnapshot();
    });

    it('Should mark name because substring at the beginning', () => {
      const newOption = createGenericOption(
        'ShortNameForTemplate',
        'ASHORTABBR',
      );
      expect(highlightOptionLabel(newOption, 'short')).toMatchSnapshot();
    });
  });

  describe('getOptionLabel', () => {
    it('Should return the label field value from an option', () => {
      const option = createGenericOption('testName');
      expect(getOptionLabel(option)).toBe(option.label);
    });
  });

  describe('getSimilarity', () => {
    const testData: [string, string | undefined, string, number][] = [
      ['name', 'abbreviation', 'test', 0],
      ['name', undefined, 'test', 0],
      ['testName', 'testAbbreviation', 'testname', 1],
      ['testName', undefined, 'testname', 1],
      ['testName', 'testAbbreviation', 'testabbreviation', 1],
      ['testName', 'testAbbreviation', 'testa', 0.3125],
      ['testName', 'testAbbreviation', 'testn', 0.625],
      ['testAb', 'testAbbreviation', 'testa', 0.8333333333333334],
      ['testAb', 'testAbbreviation', '', 0],
    ];

    it.each(testData)(
      'Should calculate similarity based on name (%s) and abbreviation (%s) by lookup value (%s) to similarity %i',
      (name, abbreviation, lookupValue, similarity) => {
        const option = createGenericOption(name, abbreviation);
        expect(getSimilarity(option, lookupValue)).toBe(similarity);
      },
    );
  });

  describe('filterOptions', () => {
    const optionA = createOption('Argon', 'Ar');
    const optionB = createOption('Gold', 'Au');

    const optionD = createOption('Copernicium', 'Cn');
    const optionC = createOption('Cobalt', 'Co');
    const optionE = createOption('CO2H');
    const optionF = createOption('Silicon', 'Si');
    const optionG = createOption('benzyl alcohol');

    it('Should filter out Gold Element', () => {
      const lookupValue = 'Argon';
      const inputArray = [optionA, optionB];
      const resultArray = [optionA];
      expect(
        filterOptions(inputArray, {
          inputValue: lookupValue,
        } as FilterOptionsState<AbbreviationOption>),
      ).toEqual(resultArray);
    });

    it('Should ignore spaces at the beginning and the end', () => {
      const lookupValue = '   Argon   ';
      const inputArray = [optionA, optionB];
      const resultArray = [optionA];
      expect(
        filterOptions(inputArray, {
          inputValue: lookupValue,
        } as FilterOptionsState<AbbreviationOption>),
      ).toEqual(resultArray);
    });

    it('Should return an empty array for empty lookup value', () => {
      const lookupValue = '';
      const inputArray = [optionB, optionA];
      const resultArray = [];
      expect(
        filterOptions(inputArray, {
          inputValue: lookupValue,
        } as FilterOptionsState<AbbreviationOption>),
      ).toEqual(resultArray);
    });

    it('Should sort by checking if name or abbreviation is starts with lookup value', () => {
      const lookupValue = 'go';
      const inputArray = [optionB, optionA];
      const resultArray = [optionB, optionA];
      expect(
        filterOptions(inputArray, {
          inputValue: lookupValue,
        } as FilterOptionsState<AbbreviationOption>),
      ).toEqual(resultArray);
    });

    it('Should correctly consider name or abbreviation while sorting', () => {
      const lookupValue = 'co';
      const inputArray = [optionF, optionG, optionD, optionC, optionE];
      const resultArray = [optionC, optionE, optionD, optionF, optionG];
      expect(
        filterOptions(inputArray, {
          inputValue: lookupValue,
        } as FilterOptionsState<AbbreviationOption>),
      ).toEqual(resultArray);
    });
  });
});
