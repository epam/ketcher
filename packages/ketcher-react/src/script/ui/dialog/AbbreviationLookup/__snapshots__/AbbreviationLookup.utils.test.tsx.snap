// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AbbreviationLookup Utils highlightOptionLabel Should ignore space at the beginning and the end 1`] = `"SHORTABBR (Very long test name)"`;

exports[`AbbreviationLookup Utils highlightOptionLabel Should mark abbreviation because it shorter (bigger similarity) 1`] = `
<React.Fragment>
  
  <mark>
    SHORT
  </mark>
  ABBR (ShortNameForTemplate)
</React.Fragment>
`;

exports[`AbbreviationLookup Utils highlightOptionLabel Should mark name because substring at the beginning 1`] = `
<React.Fragment>
  ASHORTABBR (
  <mark>
    Short
  </mark>
  NameForTemplate)
</React.Fragment>
`;

exports[`AbbreviationLookup Utils highlightOptionLabel Should not highlight if lookup value length is short 1`] = `"SHORTABBR (Very long test name)"`;

exports[`AbbreviationLookup Utils highlightOptionLabel Should return marked abbreviation 1`] = `
<React.Fragment>
  SHORT
  <mark>
    ABBR
  </mark>
   (Very long test name)
</React.Fragment>
`;

exports[`AbbreviationLookup Utils highlightOptionLabel Should return marked name 1`] = `
<React.Fragment>
  SHORTABBR (Very 
  <mark>
    long
  </mark>
   test name)
</React.Fragment>
`;
